### PR TITLE
DEV-1319 - airdrop with consent

### PR DESCRIPTION
**Pull request by grp06**
## Describe your changes
- added `airdropWithConsent`
- changed `getSignature` function on tests so that it's not hardcoded to be dependent upon a specific typehash.
- write tests for `airdropWithConsent` 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have updated the documentation accordingly. (if applicable)
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/Badges.sol [hunk 1]
```diff
@@ -270,6 +273,28 @@ contract Badges is
     return mint(msg.sender, _uri, raftTokenId);
   }
 
+  function airdropWithConsent(
+    address _recipient,
+    string calldata _uri,
+    bytes calldata _issuerSignature,
```
Comment (rahulrumalla) [inline src/Badges.sol:279]: Why not just use AgreementHash for issuer signature and RequestHash for claimant signature? Wouldn't that give us what we need? i.e., 

Issuer essential is saying "I approve this badge for this claimant"
Claimant is essentially saying "I consent to receiving this badge"

With those 2 signatures, anyone can airdrop it, no?

File: src/Badges.sol [hunk 2]
```diff
@@ -270,6 +273,28 @@ contract Badges is
...
+    bytes calldata _recipientSignature
+  ) external virtual {
+    uint256 raftTokenId = specDataHolder.getRaftTokenId(_uri);
+    require(
+      specDataHolder.isAuthorizedAdmin(raftTokenId, msg.sender),
```
Comment (rahulrumalla) [inline src/Badges.sol:284]: This doesn't match the requirements. The sender of this txn can be anyone. Please see ticket.

**Review by rahulrumalla — CHANGES_REQUESTED** [review]
A couple of things to align on..

File: src/Badges.sol (hunk 1 above)
Comment (grp06) [reply src/Badges.sol:283]: yep agreed. Updated

File: src/Badges.sol [hunk 3]
```diff
@@ -330,13 +370,9 @@ contract Badges is
...
   }
 
-  function tokenURI(uint256 _tokenId)
-    external
-    view
-    virtual
-    override
-    returns (string memory)
-  {
+  function tokenURI(
```
Comment (rahulrumalla) [inline src/Badges.sol:373]: Why did this linting stuff change? did you override some local config by any chance?

File: src/Badges.sol [hunk 4]
```diff
@@ -270,6 +269,47 @@ contract Badges is
...
 
+  /**
+   * @notice Allows minting of a badge when the caller provides valid issuer and recipient signatures
+   * @dev airdropWithConsent is called by an issuer who wants to mint a badge for a recipient with the recipient's consent.
+   * @param _recipient The address of the recipient who will receive the minted badge
+   * @param _uri The URI of the badge spec
+   * @param _issuerSignature The issuer's signature, used to verify that they approve the badge for the recipient
+   * @param _recipientSignature The recipient's signature, used to verify that they consent to receiving the badge
+   */
+  function airdropWithConsent(
```
Comment (rahulrumalla) [inline src/Badges.sol:280]:
```suggestion
  function mintWithConsent(
```

Should we call this `mintWithConsent` ? 🤔

**Review by rahulrumalla — APPROVED** [review]
Looks good!! Just check the note about naming change. Will defer to your final call on this one..

**Pull request merged**